protocols/randhound/proof.go

```diff
@@ -0,0 +1,298 @@
+func NewProof(suite abstract.Suite, g []abstract.Point, h []abstract.Point, core []ProofCore) (*Proof, error) {
+
+	if len(g) != len(h) {
+		return nil, errors.New("Received non-matching number of points")
+	}
+
+	n := len(g)
+	base := make([]ProofBase, n)
+	for i := range base {
+		base[i] = ProofBase{g: g[i], h: h[i]}
+	}
+
+	return &Proof{suite: suite, base: base, core: core}, nil
+}
+
+// Setup initializes the proof by randomly selecting a commitment v,
+// determining the challenge c = H(xG,xH,vG,vH), and the response r = v - cx.
+func (p *Proof) Setup(scalar ...abstract.Scalar) ([]abstract.Point, []abstract.Point, []ProofCore, error) {
+
+	if len(scalar) != len(p.base) {
+		return nil, nil, nil, errors.New("Received unexpected number of scalars")
+	}
+
+	n := len(scalar)
+	p.core = make([]ProofCore, n)
+	xG := make([]abstract.Point, n)
+	xH := make([]abstract.Point, n)
+	for i, x := range scalar {
+
+		xG[i] = p.suite.Point().Mul(p.base[i].g, x)
+		xH[i] = p.suite.Point().Mul(p.base[i].h, x)
+
+		// Commitment
+		v := p.suite.Scalar().Pick(random.Stream)
+		vG := p.suite.Point().Mul(p.base[i].g, v)
+		vH := p.suite.Point().Mul(p.base[i].h, v)
+
+		// Challenge
+		cb, err := crypto.HashArgsSuite(p.suite, xG[i], xH[i], vG, vH)
+		if err != nil {
+			return nil, nil, nil, err
+		}
+		c := p.suite.Scalar().Pick(p.suite.Cipher(cb))
+
+		// Response
+		r := p.suite.Scalar()
+		r.Mul(x, c).Sub(v, r)
+
+		p.core[i] = ProofCore{c, r, vG, vH}
+	}
+
+	return xG, xH, p.core, nil
+}
+
+// SetupCollective is similar to Setup with the difference that the challenge
+// is computed as the hash over all base points and commitments.
+func (p *Proof) SetupCollective(scalar ...abstract.Scalar) ([]abstract.Point, []abstract.Point, []ProofCore, error) {
+
+	if len(scalar) != len(p.base) {
+		return nil, nil, nil, errors.New("Received unexpected number of scalars")
+	}
+
+	n := len(scalar)
+	p.core = make([]ProofCore, n)
+	v := make([]abstract.Scalar, n)
+	xG := make([]abstract.Point, n)
+	xH := make([]abstract.Point, n)
+	vG := make([]abstract.Point, n)
+	vH := make([]abstract.Point, n)
+	for i, x := range scalar {
+
+		xG[i] = p.suite.Point().Mul(p.base[i].g, x)
+		xH[i] = p.suite.Point().Mul(p.base[i].h, x)
+
+		// Commitments
+		v[i] = p.suite.Scalar().Pick(random.Stream)
+		vG[i] = p.suite.Point().Mul(p.base[i].g, v[i])
+		vH[i] = p.suite.Point().Mul(p.base[i].h, v[i])
+	}
+
+	// Collective challenge
+	cb, err := crypto.HashArgsSuite(p.suite, xG, xH, vG, vH)
+	if err != nil {
+		return nil, nil, nil, err
+	}
+	c := p.suite.Scalar().Pick(p.suite.Cipher(cb))
+
+	// Responses
+	for i, x := range scalar {
+		r := p.suite.Scalar()
+		r.Mul(x, c).Sub(v[i], r)
+		p.core[i] = ProofCore{c, r, vG[i], vH[i]}
+	}
+
+	return xG, xH, p.core, nil
+}
+
+// Verify validates the proof against the given input by checking that
+// vG == rG + c(xG) and vH == rH + c(xH).
+func (p *Proof) Verify(xG []abstract.Point, xH []abstract.Point) ([]int, []int, error) {
```

Comment:
Yes good point, I'll fix the comments! Since we are doing threshold crypto here, it is allowed that some proofs fail, which is also the reason for the `good` and `bad` return values.